### DISCUSSION TITLE
EffectPass.setDepthTexture from DepthSavePass

### DISCUSSION THREAD

**Discussion by swingingtom** [pmndrs/postprocessing#331] [Q&A] [answered]
Hi postprocessing community,

Im trying to reproduce the codesandbox  [postprocessing-god-rays](https://codesandbox.io/s/qxqff?file=/src/App.js) but using `DepthSavePass` instead of `DepthPass`

Here is a sandbox showing where Im stuck :  https://codesandbox.io/s/setdepthtexture-from-depthsavepass-39zp6?file=/src/App.js

It can show/hide the saved depth texture as `TextureEffect` from the gui. Where it seems that `SaveDepthPass` is correctly saved.

But 
```
  effectPass.setDepthTexture(
    savedDepthPass.renderTarget.texture,
    RGBADepthPacking
  );
```
Seems to do absolutely nothing when feeded with `DepthSavePass` instead of `DepthPass`.
I can omit or comment out that statement, and its still working.

Working until ... another pass ( `RenderPass` or `ClearPass` ) comes after the `DepthSavePass`.
A render pass can also be enable/disabled from the gui.

Any advices are welcome. Thanks

**Comment by vanruesc** [comment, marked as answer]
It's possible to use the `DepthSavePass` in combination with a `RenderPass` to save a custom depth texture before rendering the main scene. Here's an example: https://codesandbox.io/s/postprocessing-god-rays-forked-tsq18?file=/src/App.js

This setup unearthed a problem though. The `DepthSavePass` uses `RGBADepthPacking` by default, but this produces fully ~white~ black pixels for fragments that read ~0~ 1 in the input depth texture. I think this might be a quirk of the RGBA packing function, but I'm not sure if and how that could be fixed. Here's a comparison of the discrepancy:

| DepthSavePass | DepthPass |
|-|-|
| <img width="230" src="https://user-images.githubusercontent.com/9214917/151081207-0bac6b52-7703-4493-b094-52b2e887e83a.jpg"> | <img width="230" src="https://user-images.githubusercontent.com/9214917/151081211-6bba254d-4c1e-48b2-aab0-f21b37ef71b0.jpg"> |

`BasicDepthPacking` can be used instead, but this will store depth in a `FloatType` texture which may not be supported on some devices.

> Seems to do absolutely nothing when feeded with DepthSavePass instead of DepthPass.
I can omit or comment out that statement, and its still working.

Setting a custom depth texture must be done after the pass was added to the composer. Otherwise the depth texture will be reset.

From [#330](https://github.com/vanruesc/postprocessing/issues/330#issuecomment-1020878126):

> But is there any other differences? On the internal structure or something else? I tried to replace DepthPass with DepthSavePass in godrays example as explained here #331 but something works different, and I struggle to find what.

The `DepthSavePass` copies the contents of a depth texture into another render target. It doesn't produce depth by itself and requires a depth texture as input which can be produced by a `DepthPass` or as a by-product of a `RenderPass` with a [DepthTexture](https://threejs.org/docs/?q=deptht#api/en/textures/DepthTexture) on the output render target.

**Reply by swingingtom** [reply]
Thanks you for the provided informations and the sandbox. 
I was able to notice the behaviours you described in your post by messing with the sandbox. Much more clear now !
